File: graph/1717.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <math.h>
#define fastio cin.tie(0)->sync_with_stdio(0)
#define ll long long
using namespace std;


int find(int x, vector<int>& parent) {
    if (parent[x] == x) return x;
    return parent[x] = find(parent[x], parent);
}

void merge(int x, int y, vector<int>& parent) {
    x = find(x, parent);
    y = find(y, parent);
    if (x == y) return;
    parent[y] = x;
}


int main()
{
    fastio;
    vector<string> ans;
    int n, m;
    cin >> n >> m;
    vector<int> parent(n + 1, 0);
    for (int i = 1; i <= n; i++)
        parent[i] = i;
    int t, x, y;
    for (int i = 0; i < m; i++) {
        cin >> t >> x >> y;
        if (t == 0/*union*/) {
            merge(x, y, parent);
        }
        else if(t==1){ /*find*/
            int parent_x = find(x, parent);
            int parent_y = find(y, parent);
            if (parent_x == parent_y/*has same parent*/) {
                ans.push_back("YES");
            }
            else {
                ans.push_back("NO");
            }
        }
    }
    for (int i = 0; i < ans.size(); i++) {
        cout << ans[i] << '\n';
    }
    return 0;
}

File: graph/1717.java
import java.util.Scanner;

public class Main {
    static int[] parent;

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int m = sc.nextInt();
        parent = new int[n + 1];
        for (int i = 1; i <= n; i++)
            parent[i] = i;
        for (int i = 0; i < m; i++) {
            int type = sc.nextInt();
            int ele1 = sc.nextInt();
            int ele2 = sc.nextInt();
            if (type == 0) {
                // union
                union(ele1, ele2);
            } else {
                if (isSame(ele1, ele2)) {
                    System.out.println("YES");
                } else {
                    System.out.println("NO");
                }
            }
        }
    }

    public static void union(int a, int b) {
        a = find(a);
        b = find(b);
        if (a != b) {
            parent[b] = a;
        }
    }

    public static int find(int a) {
        if (a == parent[a])
            return a;
        else {
            return parent[a] = find(parent[a]);
        }
    }

    public static boolean isSame(int a, int b) {
        a = find(a);
        b = find(b);
        if (a == b) {
            return true;
        } else {
            return false;
        }
    }
}
